"for... in" should not be used with arrays. This causes bugs!

agamemnus opened the pull request:
See: https://github.com/katspaugh/wavesurfer.js/issues/1014.

I also changed "entry" to "canvas" since they are canvases. And, uh, just a few styling ideas.

katspaugh commented:
Could you just replace the `for-in` with a `forEach` without the other stylistic changes? Btw, `forEach` takes a third argument which is the context of the callback, so you could pass `this` and not introduce another `var my = this` to keep the context.

agamemnus commented:
I already ended up removing many of those changes due to your draconian style checker! The rest is as follows --

There are only four differences now, other than the forEach fixes, and corresponding "my" variables -- the "my" variables isn't something I came up with as it was present in an existing function.
1) Some brackets had an errant starting space: e.g.: `{ width: elementWidth + 'px'}`. and the `getImage` function didn't have a space between it and the parameters.
2) `entry` -> `canvas` and `lastEntry` -> `lastCanvas`. IMO, if you have `this.canvases`, then one of them should be a `canvas` and not an `entry`.
3) Some breaking up of a multi-line var:

````
var elementWidth = Math.round(width / this.params.pixelRatio),
totalWidth = Math.round(this.width / this.params.pixelRatio);
````
to
````
var elementWidth = Math.round(width / this.params.pixelRatio);
var totalWidth   = Math.round(this.width / this.params.pixelRatio);
````
and
````
var entry = {},
leftOffset = this.maxCanvasElementWidth * this.canvases.length;
````
to
````
var canvas = {};
var leftOffset = this.maxCanvasElementWidth * this.canvases.length;
````
4)
````
var endCanvas = Math.min(Math.ceil((x + width) / this.maxCanvasWidth) + 1,
this.canvases.length);
````
to
````
var endCanvas = Math.min(Math.ceil((x + width) / this.maxCanvasWidth) + 1, this.canvases.length);
````

katspaugh commented:
While some of your changes make sense (I'll mark them with a thumbs-up in the changeset), some are arbitrary, and I would rather not create more maintenance work for the `next` branch.

katspaugh commented on src/drawer.multicanvas.js:
```diff
@@ -52,24 +53,24 @@ WaveSurfer.util.extend(WaveSurfer.Drawer.MultiCanvas, {
         }
 
-        for (var i in this.canvases) {
+        my.canvases.forEach (function (canvas, i) {
```
👍  except for the extra space after `forEach` – it's not needed.

katspaugh commented on src/drawer.multicanvas.js:
```diff
@@ -91,51 +92,52 @@ WaveSurfer.util.extend(WaveSurfer.Drawer.MultiCanvas, {
+        var my = this;
+        my.canvases.forEach (function (canvas) {
+            my.clearWaveForEntry(canvas);
+        });
```
👍 

```
this.canvases.forEach(function (canvas) {
    this.clearWaveForEntry(canvas);
}, this);
```

katspaugh commented on src/drawer.multicanvas.js:
```diff
@@ -223,31 +223,29 @@ WaveSurfer.util.extend(WaveSurfer.Drawer.MultiCanvas, {
-            this.drawLineToContext(entry, entry.progressCtx, peaks, absmax, halfH, offsetY, start, end);
-        }
+        var my = this;
+        my.canvases.forEach (function (canvas) {
```
The same comment as above.

katspaugh commented on src/drawer.multicanvas.js:
```diff
@@ -333,16 +330,16 @@ WaveSurfer.util.extend(WaveSurfer.Drawer.MultiCanvas, {
      *
      */
-    getImage: function(type, quality) {
+    getImage: function (type, quality) {
```
:thumbsup:

agamemnus commented:
OK, give me a few minutes to make the appropriate changes.

agamemnus commented:
Ok, I removed almost all of the other changes.

I also implemented your suggestion about inserting a `this` variable in forEach.

katspaugh commented:
I still see a lot of other changes in the diff.

agamemnus commented:
I fixed some errors.

The only other change now is removing space between opening/closing brackets and text. You don't really even have that consistent...

agamemnus commented:
Ok, a few more I didn't notice.

agamemnus commented:
I really hate those extraneous spaces, but I added them back. I also added some extra extraneous spaces in style definition calls for it to be consistent.

katspaugh commented on src/drawer.multicanvas.js:
```diff
@@ -191,23 +190,22 @@ WaveSurfer.util.extend(WaveSurfer.Drawer.MultiCanvas, {
 
         // Support arrays without negative peaks
-        var hasMinValues = [].some.call(peaks, function (val) { return val < 0; });
+        var hasMinValues = [].some.call(peaks, function (val) {return val < 0;});
```
Could you revert this change, too?

agamemnus commented:
Ok.

katspaugh commented:
Merged, thanks!
